From 3a47b479f73c236c4fc928f7961a1103c0ea9a65 Mon Sep 17 00:00:00 2001
From: Ian Dannapel <ian.dannapel@iris-sensing.com>
Date: Fri, 13 May 2022 15:27:16 +0200
Subject: [PATCH] imx8mp-evk.dts: Add epc660 and tc358746 nodes to dts

- Remove dtb from Makefile which depends on the previous ov5640 camera
node being present in the -evk devicetree
- Set tc358746 clock to non-continuous mode
- Set hs-settle accordingly to the 216Mhz CSI Clock frequency
- Disable VPU/GPU
---
 arch/arm64/boot/dts/freescale/Makefile       |  3 +-
 arch/arm64/boot/dts/freescale/imx8mp-evk.dts | 94 ++++++++++++++------
 2 files changed, 70 insertions(+), 27 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index b2729f9a2f1a..96166fd652fc 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -128,7 +128,7 @@ dtb-$(CONFIG_ARCH_MXC) += imx8mp-evk.dtb imx8mp-evk-rm67191.dtb imx8mp-evk-it626
 			  imx8mp-evk-basler-ov2775.dtb imx8mp-evk-dual-basler.dtb \
 			  imx8mp-evk-dual-ov2775.dtb imx8mp-evk-spdif-lb.dtb imx8mp-evk-dsp.dtb \
 			  imx8mp-evk-sof-wm8960.dtb imx8mp-evk-sof-pdm.dtb \
-			  imx8mp-evk-os08a20-ov5640.dtb imx8mp-evk-os08a20.dtb \
+			  imx8mp-evk-os08a20.dtb \
 			  imx8mp-evk-dual-os08a20.dtb \
 			  imx8mp-evk-iqaudio-dacplus.dtb imx8mp-evk-iqaudio-dacpro.dtb imx8mp-evk-hifiberry-dacplus.dtb \
 			  imx8mp-evk-hifiberry-dac2.dtb imx8mp-evk-hifiberry-dacplusadc.dtb \
@@ -157,7 +157,6 @@ imx8mp-evk-revb4-basler-ov2775-dtbs := imx8mp-evk-basler-ov2775.dtb imx8mp-evk-r
 imx8mp-evk-revb4-dual-basler-dtbs := imx8mp-evk-dual-basler.dtb imx8mp-evk-revb4.dtbo
 imx8mp-evk-revb4-dual-ov2775-dtbs := imx8mp-evk-dual-ov2775.dtb imx8mp-evk-revb4.dtbo
 imx8mp-evk-revb4-spdif-lb-dtbs := imx8mp-evk-spdif-lb.dtb imx8mp-evk-revb4.dtbo
-imx8mp-evk-revb4-os08a20-ov5640-dtbs := imx8mp-evk-os08a20-ov5640.dtb imx8mp-evk-revb4.dtbo
 imx8mp-evk-revb4-os08a20-dtbs := imx8mp-evk-os08a20.dtb imx8mp-evk-revb4.dtbo
 imx8mp-evk-revb4-dual-os08a20-dtbs := imx8mp-evk-dual-os08a20.dtb imx8mp-evk-revb4.dtbo
 imx8mp-evk-revb4-iqaudio-dacplus-dtbs := imx8mp-evk-iqaudio-dacplus.dtb imx8mp-evk-revb4.dtbo
diff --git a/arch/arm64/boot/dts/freescale/imx8mp-evk.dts b/arch/arm64/boot/dts/freescale/imx8mp-evk.dts
index 695d76afdfaf..65de733ee5c1 100644
--- a/arch/arm64/boot/dts/freescale/imx8mp-evk.dts
+++ b/arch/arm64/boot/dts/freescale/imx8mp-evk.dts
@@ -221,6 +221,13 @@ usb3_data_ss: endpoint {
 			};
 		};
 	};
+
+	clk_serialize: clock_serialize {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <32000000>;
+		clock-output-names = "32m_serialize";
+	};
 };
 
 &flexspi {
@@ -560,25 +567,60 @@ it6263_in: endpoint {
 		};
 	};
 
-	ov5640_0: ov5640_mipi@3c {
-		compatible = "ovti,ov5640";
-		reg = <0x3c>;
+	epc660_mipi: epc660_mipi@22 {
+		compatible = "epc660";
+		reg = <0x22>;
+		status = "okay";
 		pinctrl-names = "default";
-		pinctrl-0 = <&pinctrl_csi0_pwn>, <&pinctrl_csi0_rst>, <&pinctrl_csi_mclk>;
-		clocks = <&clk IMX8MP_CLK_IPP_DO_CLKO2>;
-		clock-names = "xclk";
-		assigned-clocks = <&clk IMX8MP_CLK_IPP_DO_CLKO2>;
-		assigned-clock-parents = <&clk IMX8MP_CLK_24M>;
-		assigned-clock-rates = <24000000>;
-		powerdown-gpios = <&gpio2 11 GPIO_ACTIVE_HIGH>;
+		pinctrl-0 = <&pinctrl_csi0_rst>, <&pinctrl_csi_mclk>;
 		reset-gpios = <&gpio1 6 GPIO_ACTIVE_LOW>;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		port@0 {
+			reg = <0>;
+			epc660_parallel_out: endpoint {
+				bus-type = <5>;
+				bus-width = <12>;
+				hsync-active = <0>;
+				vsync-active = <0>;
+				field-even-active = <0>;
+				pclk-sample = <1>;
+				remote-endpoint = <&tc358748_parallel_in>;
+			};
+		};
+	};
+
+	tc358748_adapter: csi-bridge@e {
+		compatible = "toshiba,tc358746";
+		reg = <0x0e>;
 		status = "okay";
 
-		port {
-			ov5640_mipi_0_ep: endpoint {
-				remote-endpoint = <&mipi_csi0_ep>;
+		clocks = <&clk_serialize>;
+		default-input = <0>;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		port@0 {
+			reg = <0>;
+			tc358748_parallel_in: endpoint {
+				bus-type = <5>;
+				bus-width = <12>;
+				hsync-active = <0>;
+				vsync-active = <0>;
+				field-even-active = <0>;
+				remote-endpoint = <&epc660_parallel_out>;
+			};
+		};
+		port@1 {
+			reg = <1>;
+			tc358748_mipi_out: endpoint {
 				data-lanes = <1 2>;
-				clock-lanes = <0>;
+				clock-noncontinuous;
+				link-frequencies = /bits/ 64 <216000000>;
+				remote-endpoint = <&mipi_csi0_ep>;
 			};
 		};
 	};
@@ -1389,35 +1431,35 @@ MX8MP_IOMUXC_GPIO1_IO15__CCM_CLKO2	0x50
 };
 
 &vpu_g1 {
-	status = "okay";
+	status = "disabled";
 };
 
 &vpu_g2 {
-	status = "okay";
+	status = "disabled";
 };
 
 &vpu_vc8000e {
-	status = "okay";
+	status = "disabled";
 };
 
 &vpu_v4l2 {
-	status = "okay";
+	status = "disabled";
 };
 
 &gpu_3d {
-	status = "okay";
+	status = "disabled";
 };
 
 &gpu_2d {
-	status = "okay";
+	status = "disabled";
 };
 
 &ml_vipsi {
-	status = "okay";
+	status = "disabled";
 };
 
 &mix_gpu_ml {
-	status = "okay";
+	status = "disabled";
 };
 
 &mipi_csi_0 {
@@ -1425,11 +1467,13 @@ &mipi_csi_0 {
 
 	port {
 		mipi_csi0_ep: endpoint {
-			remote-endpoint = <&ov5640_mipi_0_ep>;
 			data-lanes = <2>;
-			csis-hs-settle = <13>;
-			csis-clk-settle = <2>;
+			clock-lanes = <1>;
+			remote-endpoint = <&tc358748_mipi_out>;
+			csis-hs-settle = <4>;
+			csis-clk-settle = <0>;
 			csis-wclk;
+			link-frequencies = /bits/ 64 <216000000>;
 		};
 	};
 };
-- 
2.47.0

